dbms/src/Functions/FunctionsDateTime.h

```diff
@@ -2814,42 +2813,282 @@ class FunctionExtractMyDateTime : public IFunction
+            if (datetime.hour == duration.hours() && datetime.minute == duration.minutes() && datetime.second == duration.seconds() && datetime.year > 0)
             {
-                constantDatetime<F>(from->getUInt(0), from->size(), vec_to);
+                return ExtractMyDateTimeImpl::extractDayMinute(datetime.toPackedUInt());
             }
-            else
+        }
+        return result;
+    }
+
+    static Int64 extractDayHour(String dtStr)
+    {
+        Field duration_field = parseMyDuration(dtStr);
+        Int64 result = 0;
+        if (duration_field.isNull())
+        {
+            // TODO: should return null
+            return 0;
+        }
+        MyDuration duration(duration_field.template safeGet<Int64>());
+        result = ExtractMyDurationImpl::extractDayHour(duration.nanoSecond());
+
+        Field datetime_field = parseMyDateTime(dtStr);
+        if (!datetime_field.isNull())
+        {
+            MyDateTime datetime(datetime_field.template safeGet<UInt64>());
+            if (datetime.hour == duration.hours() && datetime.minute == duration.minutes() && datetime.second == duration.seconds() && datetime.year > 0)
             {
-                vectorDatetime<F>(data, vec_to);
+                return ExtractMyDateTimeImpl::extractDayHour(datetime.toPackedUInt());
             }
         }
+        return result;
+    }
+};
+
+class FunctionExtractMyDateTimeFromString : public IFunction
+{
+public:
+    static constexpr auto name = "extractMyDateTimeFromString";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionExtractMyDateTimeFromString>(); };
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (!arguments[0]->isString())
+            throw TiFlashException(fmt::format("First argument for function {} (unit) must be String", getName()), Errors::Coprocessor::BadRequest);
+
+        if (!arguments[1]->isString())
+            throw TiFlashException(
+                fmt::format("Illegal type {} of second argument of function {}. Must be String.", arguments[1]->getName(), getName()),
+                Errors::Coprocessor::BadRequest);
+
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {0}; }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        const auto * unit_column = checkAndGetColumnConst<ColumnString>(block.getByPosition(arguments[0]).column.get());
+        if (!unit_column)
+            throw TiFlashException(
+                fmt::format("First argument for function {} must be constant String", getName()),
+                Errors::Coprocessor::BadRequest);
```

Comment:
throw Exception(xx, ErrorCodes::ILLEGAL_COLUMN)?

Reply:
this is a constant string _unit_, does not represent a column

Reply:
I will change it to ILLEGAL_TYPE_OF_ARGUMENT